src/wasm_opt.rs

```diff
@@ -0,0 +1,356 @@
+    * Debian/Ubuntu: apt-get install binaryen\n\
+    * Homebrew: brew install binaryen\n\
+    * Arch Linux: pacman -S binaryen\n\
+    * Windows: binary releases at https://github.com/WebAssembly/binaryen/releases";
+
+/// A helpful struct for interacting with Binaryen's `wasm-opt` tool.
+pub struct WasmOptHandler {
+    /// The path to the `wasm-opt` binary.
+    wasm_opt_path: PathBuf,
+    /// The optimization level that should be used when optimizing the Wasm binary.
+    optimization_level: OptimizationPasses,
+    /// Whether or not to keep debugging information in the final Wasm binary.
+    keep_debug_symbols: bool,
+    /// The version number of the `wasm-opt` binary being executed.
+    _version: u32,
+}
+
+impl WasmOptHandler {
+    /// Generate a new instance of the handler.
+    ///
+    /// Fails if the `wasm-opt` binary is not installed on the system, or if an outdated `wasm-opt`
+    /// binary is used (currently a version >= 99 is required).
+    pub fn new(
+        optimization_level: OptimizationPasses,
+        keep_debug_symbols: bool,
+    ) -> Result<Self> {
+        let which = which::which("wasm-opt");
+        if which.is_err() {
+            anyhow::bail!(WASM_OPT_INSTALLATION_SUGGESTION.to_string().bright_yellow());
+        }
+
+        let wasm_opt_path =
+            which.expect("we just checked if `which` returned an err; qed");
+        tracing::info!("Path to wasm-opt executable: {}", wasm_opt_path.display());
+
+        let version =
+            Self::check_wasm_opt_version_compatibility(wasm_opt_path.as_path())?;
+
+        Ok(Self {
+            wasm_opt_path,
+            optimization_level,
+            keep_debug_symbols,
+            _version: version,
+        })
+    }
+
+    /// Attempts to perform optional Wasm optimization using Binaryen's `wasm-opt` tool.
+    ///
+    /// If successful, the optimized Wasm binary is written to `dest_wasm`.
+    pub fn optimize(
+        &self,
+        dest_wasm: &PathBuf,
+        contract_artifact_name: &String,
+    ) -> Result<OptimizationResult> {
+        // We'll create a temporary file for our optimized Wasm binary. Note that we'll later
+        // overwrite this with the original path of the Wasm binary.
+        let mut dest_optimized = dest_wasm.clone();
+        dest_optimized.set_file_name(format!("{}-opt.wasm", contract_artifact_name));
+
+        tracing::info!(
+            "Optimization level passed to wasm-opt: {}",
+            self.optimization_level
+        );
+
+        let mut command = Command::new(self.wasm_opt_path.as_path());
+        command
+        .arg(dest_wasm.as_os_str())
+        .arg(format!("-O{}", self.optimization_level))
+        .arg("-o")
+        .arg(dest_optimized.as_os_str())
+        // the memory in our module is imported, `wasm-opt` needs to be told that
+        // the memory is initialized to zeroes, otherwise it won't run the
+        // memory-packing pre-pass.
+        .arg("--zero-filled-memory");
```

Comment:
Surely this whole block should be indented?

Reply:
RustFmt seems indifferent, but yeah I've indented it